src/plugins/collision/collider_backend.rs

```diff
@@ -250,102 +247,138 @@ fn init_colliders<C: AnyCollider>(
     }
 }
 
-/// Creates [`Collider`]s from [`AsyncCollider`]s if the meshes have become available.
-#[cfg(all(
-    feature = "3d",
-    feature = "async-collider",
-    feature = "default-collider"
-))]
-pub fn init_async_colliders(
+/// Creates [`Collider`]s from [`ColliderConstructor`]s if the meshes have become available.
+fn init_collider_constructors(
     mut commands: Commands,
     meshes: Res<Assets<Mesh>>,
-    async_colliders: Query<(Entity, &Handle<Mesh>, &AsyncCollider)>,
+    constructors: Query<(
+        Entity,
+        Option<&Handle<Mesh>>,
+        Option<&Collider>,
+        Option<&Name>,
+        &ColliderConstructor,
+    )>,
 ) {
-    for (entity, mesh_handle, async_collider) in async_colliders.iter() {
-        if let Some(mesh) = meshes.get(mesh_handle) {
-            let collider = match &async_collider.0 {
-                ComputedCollider::TriMesh => Collider::trimesh_from_mesh(mesh),
-                ComputedCollider::TriMeshWithFlags(flags) => {
-                    Collider::trimesh_from_mesh_with_config(mesh, *flags)
-                }
-                ComputedCollider::ConvexHull => Collider::convex_hull_from_mesh(mesh),
-                ComputedCollider::ConvexDecomposition(params) => {
-                    Collider::convex_decomposition_from_mesh_with_config(mesh, params)
-                }
-            };
-            if let Some(collider) = collider {
-                commands
-                    .entity(entity)
-                    .insert(collider)
-                    .remove::<AsyncCollider>();
-            } else {
-                error!("Unable to generate collider from mesh {:?}", mesh);
+    for (entity, mesh_handle, existing_collider, name, constructor) in constructors.iter() {
+        let name = name.map(|n| n.as_str()).unwrap_or("<unnamed>");
+        if existing_collider.is_some() {
+            warn!(
+                "Tried to add a collider to entity {name} via {constructor:#?}, \
+                but that entity already holds a collider. Skipping.",
+            );
+            continue;
+        }
+        let mesh = if constructor.requires_mesh() {
+            let mesh_handle = mesh_handle.unwrap_or_else(|| panic!(
+                "Tried to add a collider to entity {name} via {constructor:#?} that requires a mesh, \
+                but no mesh handle was found"));
```

Comment:
Nope. I'm doing string interpolation here, so if anything it would be `expect(format!(...))`, which would call `format!` independently of whether the error even occurred, which clippy will correctly warn about. This is the right way.